helm/kyverno-policies-connectivity/templates/Pod.yaml

```diff
@@ -18,6 +18,32 @@ spec:
             - {{ . | quote }}
           {{- end }}
 {{- end }}
+      context:
+        - name: kubeadmConfig
+          configMap:
+            name: kubeadm-config
+            namespace: kube-system
+        - name: clusterConfiguration
+          variable:
+            value: {{`"{{ kubeadmConfig.data.ClusterConfiguration | parse_yaml(@) }}"`}}
+        - name: apiEndpoint
+          variable:
+            value: {{`"{{ clusterConfiguration.controlPlaneEndpoint | split(@, ':') | [0] }}"`}}
+        - name: dnsDomain
+          variable:
+            value: {{`"{{ clusterConfiguration.networking.dnsDomain }}"`}}
+        - name: podSubnet
+          variable:
+            value: {{`"{{ clusterConfiguration.networking.podSubnet }}"`}}
```

Comment:
Interestingly in openstack cluster I don't see podSubnet (serviceSubnet is there). Maybe it isn't set when it's default or something?

Reply:
I checked CAPA, CAPO and CAPVCD again. As you said, the only diff is that CAPO hasn't podSubnet in the config. Interestingly, it doesn't have that config in node specs as others have.

Reply:
In CAPA and CAPVCD, for cilium configuration, we use
```
ipam:
  mode: kubernetes
```

In CAPO, we use
```
ipam:
  mode: cluster-pool
  operator:
    clusterPoolIPv4MaskSize: 24
    clusterPoolIPv4PodCIDR: 10.0.0.0/8
    clusterPoolIPv4PodCIDRList: []
    clusterPoolIPv6MaskSize: 120
    clusterPoolIPv6PodCIDR: fd00::/104
    clusterPoolIPv6PodCIDRList: []
```

That is the root of the diff.

Reply:
Also, for CAPVCD, we set podCIDR in cluster spec explicitly.
https://github.com/giantswarm/cluster-cloud-director/blob/main/helm/cluster-cloud-director/templates/cluster.yaml#L16

However, it is not used actively. For example, in guppy,

This is cluster's spec.
```
      pods:
        cidrBlocks:
        - 100.96.0.0/11
```

This is the value in kubeadm config.
```
    networking:
      podSubnet: 100.96.0.0/11
```

This is node spec
```
spec:
  podCIDR: 100.96.3.0/24
  podCIDRs:
  - 100.96.3.0/24
```

but this is cilium annotation in Node CRs.
```
➜ k get node -A -o yaml |grep io.cilium.network.ipv4-pod-cidr
      io.cilium.network.ipv4-pod-cidr: 10.0.3.0/24
      io.cilium.network.ipv4-pod-cidr: 10.0.5.0/24
      io.cilium.network.ipv4-pod-cidr: 10.0.0.0/24
      io.cilium.network.ipv4-pod-cidr: 10.0.2.0/24
      io.cilium.network.ipv4-pod-cidr: 10.0.1.0/24
      io.cilium.network.ipv4-pod-cidr: 10.0.4.0/24
```

Reply:
I found that there is an inconsistency in `guppy` (CAPVCD cluster).  Its ipam mode was wrong during the first deployment of the cilium. https://github.com/giantswarm/mc-bootstrap/pull/360/files

In CAPVCD clusters, we will use ipam.mode=kubernetes and the values in kubeadm-config will be reliable.

For CAPO, we use ipam.mode=cluster-pool and kyverno fails with this error
```
E1212 13:14:38.296168       1 mutation.go:76] EngineMutate "msg"="failed to load context" "error"="failed to substitute variables in context entry podSubnet &JSON{Raw:*[34 123 123 32 99 108 117 115 116 101 114 67 111 110 102 105 103 117 114 97 116 105 111 110 46 110 101 116 119 111 114 107 105 110 103 46 112 111 100 83 117 98 110 101 116 32 125 125 34],}: failed to resolve clusterConfiguration.networking.podSubnet at path : JMESPath query failed: Unknown key \"podSubnet\" in path" "kind"="Pod" "name"="erkan-test" "namespace"="org-giantswarm" "policy"="erkan-test-policy" "rule"="erkan-test-rule"
```

On the other hand, we don't support proxy configuration for CAPO yet.

@bavarianbidi @kopiczko what do you think? can we rely on the cilium config at this point?

Reply:
Agreed. CAPO proxy issue https://github.com/giantswarm/roadmap/issues/1783